engine/eConstr.ml

```diff
@@ -552,32 +552,116 @@ let compare_gen k eq_inst eq_sort eq_constr c1 c2 =
+  | Declarations.Monomorphic_ind _ ->
+    assert (Univ.Instance.length u1 = 0 && Univ.Instance.length u2 = 0);
+    cstrs
+  | Declarations.Polymorphic_ind _ ->
+     enforce_eq_instances_univs true u1 u2 cstrs
+  | Declarations.Cumulative_ind cumi ->
+    let num_param_arity =
+      mind.Declarations.mind_nparams +
+      mind.Declarations.mind_packets.(ind).Declarations.mind_nrealargs
+    in
+    if not (Int.equal num_param_arity nargs) then
+      enforce_eq_instances_univs true u1 u2 cstrs
+    else
+      let newcstrs = cmp_cumulative_inductive_instances cv_pb cumi u1 u2 in
+      let newcstrs = of_constraints true newcstrs in
+      Constraints.union cstrs newcstrs
+
+let cmp_constructors (mind, ind, cns) nargs u1 u2 cstrs =
+  let open Universes in
+  match mind.Declarations.mind_universes with
+  | Declarations.Monomorphic_ind _ ->
+    assert (Univ.Instance.length u1 = 0 && Univ.Instance.length u2 = 0);
+    cstrs
+  | Declarations.Polymorphic_ind _ ->
+    enforce_eq_instances_univs true u1 u2 cstrs
+  | Declarations.Cumulative_ind cumi ->
+    let num_cnstr_args =
+      let nparamsctxt =
+        mind.Declarations.mind_nparams +
+        mind.Declarations.mind_packets.(ind).Declarations.mind_nrealargs
+        (* Context.Rel.length mind.Declarations.mind_params_ctxt *) in
+      nparamsctxt + mind.Declarations.mind_packets.(ind).Declarations.mind_consnrealargs.(cns - 1)
+    in
+    if not (Int.equal num_cnstr_args nargs) then
+      enforce_eq_instances_univs true u1 u2 cstrs
+    else
+      let newcstrs = cmp_cumulative_inductive_instances Reduction.CONV cumi u1 u2 in
+      let newcstrs = of_constraints true newcstrs in
+      Constraints.union cstrs newcstrs
+
+let eq_universes env sigma cv_pb cstrs ref nargs l l' =
+  let open Universes in
+  let l = EInstance.kind sigma (EInstance.make l) in
+  let l' = EInstance.kind sigma (EInstance.make l') in
+  match ref with
+  | VarRef _ -> assert false (* variables don't have instances *)
+  | ConstRef _ ->
+    cstrs := enforce_eq_instances_univs true l l' !cstrs; true
+  | IndRef ind ->
+    let mind = Environ.lookup_mind (fst ind) env in
+    cstrs := cmp_inductives cv_pb (mind,snd ind) nargs l l' !cstrs;
+    true
+  | ConstructRef ((mi,ind),ctor) ->
+    let mind = Environ.lookup_mind mi env in
+    cstrs := cmp_constructors (mind,ind,ctor) nargs l l' !cstrs;
+    true
+
+let test_constr_universes env sigma leq m n =
   let open Universes in
   let kind c = kind_upto sigma c in
   if m == n then Some Constraints.empty
-  else 
+  else
+    let cv_pb = if leq then Reduction.CUMUL else Reduction.CONV in
     let cstrs = ref Constraints.empty in
-    let eq_universes strict l l' = 
-      let l = EInstance.kind sigma (EInstance.make l) in
-      let l' = EInstance.kind sigma (EInstance.make l') in
-      cstrs := enforce_eq_instances_univs strict l l' !cstrs; true in
+    let eq_universes ref nargs l l' = eq_universes env sigma cv_pb cstrs ref nargs l l' in
```

Comment:
This is bugged, when the overall function is called in the CUMUL case this gets used with CUMUL even on subterms where it should be CONV.
Will fix after messing with some attempts to improve perf.